chore(WD-24270): SOLID Design Principles Examples and Implementation

muhammad-ali-pk opened the pull request:
## Done

 - Implemented SOLID design principles
 - Refactored the codebase to implement each of these principles (where applicable)
 - (Drive by): Added an in-memory mock server as part of [Liskov Substitution Principle (LSP) principle](https://github.com/canonical/cs.canonical.com/pull/240/commits/e443a98a1a3847144bb4222911d2eccc050988b2)
 - (Drive by fixes): 
     - [fix: first asset of 20years page on ubuntu showing wrong file-type](https://github.com/canonical/cs.canonical.com/pull/240/commits/2a8f3222f5c7e8b507d101a273816ae06c9b93d2)
     - [chore: open tree view and highlight new page, on page creation](https://github.com/canonical/cs.canonical.com/pull/240/commits/ae745aabaa1c2aad86e7326ffa8177d1daa7f210)
     - [fix: create page request not using provided copydoc](https://github.com/canonical/cs.canonical.com/pull/240/commits/cd09e80486b96952080dd46035ffe4011be6a488)


## SOLID Design Principles
Set of principles that make your code more:

- Maintainable
- Scalable
- Flexible
- Understandable
- Refactorable

These are

1. Single Responsibility Principle (SRP)
2. Open/Closed Principle (OCP)
3. Liskov Substitution Principle (LSP)
4. Interface Segregation Principle (ISP)
5. Dependency Inversion Principle (DIP)

### Single Responsibility Principle (SRP)

A class/program/component should have only one reason to change, meaning it should have only one job or responsibility.
If a class/program/component does too many things, changes in one responsibility may break others. Keep them focused.

See 🔗 [`Single Responsibility Principle`](https://github.com/canonical/cs.canonical.com/pull/240/commits/b09479ad40d0c7824c378dd1a0418fc30799120a) for the implementation examples.

### Open/Closed Principle (OCP)

Software entities (classes, modules, functions, components) should be open for extension but closed for modification.
You should be able to add new functionality without changing existing code, avoiding regression bugs.

See 🔗 [`Open/Closed Principle (OCP)`](https://github.com/canonical/cs.canonical.com/pull/240/commits/2bac4711d6dd869c2c1ab93c981570f456a38be4) for the implementation examples.

### Liskov Substitution Principle (LSP)

Subclasses should be substitutable for their base classes without breaking the program.
If B is a subclass of A, then objects of A should be replaceable with objects of B without altering the desirable properties of the program.

See 🔗 [`Liskov Substitution Principle (LSP)`](https://github.com/canonical/cs.canonical.com/pull/240/commits/e443a98a1a3847144bb4222911d2eccc050988b2) for the implementation examples.

### Interface Segregation Principle (ISP)

Clients should not be forced to depend on interfaces they do not use.
Instead of one fat interface with many methods, create smaller, specific interfaces.
This could be applied to Abstractions, Interfaces, Functional Props and more.

See 🔗 [`Interface Segregation Principle`](https://github.com/canonical/cs.canonical.com/pull/240/commits/535ab4c247e4b8371479790d8d538ed792a14bbe) for the implementation examples.

### Dependency Inversion Principle (DIP)

High-level modules should not depend on low-level modules; both should depend on abstractions.
Instead of directly using concrete classes, depend on interfaces/abstract classes so you can swap implementations easily.

See 🔗 [`Dependency Inversion Principle`](https://github.com/canonical/cs.canonical.com/pull/240/commits/c63f794b6371562e3559d073434ae22de80f9a14) for the implementation examples.




## QA

### QA steps

 - Clone the repository
 - Run the project

```bash
docker run -d -p 5432:5432 -e POSTGRES_PASSWORD=postgres postgres
docker run -d -p 6379:6379 redis
dotrun
```

In another terminal
```bash
yarn dev
```

- Verify the application works as expected
- Verify the [Playwright Tests / run-playwright (pull_request)](https://github.com/canonical/cs.canonical.com/actions/runs/17975454039/job/51127766599?pr=240) action is successful

## Fixes

 - Fixes [WD-24270](https://warthogs.atlassian.net/browse/WD-24270), [WD-27693](https://warthogs.atlassian.net/browse/WD-27693), [WD-27694](https://warthogs.atlassian.net/browse/WD-27694), [WD-27704](https://warthogs.atlassian.net/browse/WD-27704)


[WD-24270]: https://warthogs.atlassian.net/browse/WD-24270?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://cs-canonical-com-240.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/23856)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

akbarkz commented on static/client/components/WebpageAssets/index.tsx:
```diff
@@ -5,11 +5,15 @@ import { Pagination, Spinner } from "@canonical/react-components";
-import type { IPage } from "@/services/api/types/pages";
 
-const WebpageAssets: React.FC<{ page: IPage }> = ({ page }) => {
+interface WebpageAssetsProps {
```
why not follow the same approach everywhere and move this to a separate types.ts file?

muhammad-ali-pk replied:
In my opinion, if a type is only used in one place (like WebpageAssetsProps type that's only gonna be used in WebpageAssets/index.ts component), it's better to declare it in the same file. This will help keeping the types files count in check. We can always move it to dedicated types file, if it's required to be used in another place.
Wdyt?

akbarkz commented on static/client/hooks/useUsersRequest/index.ts:
```diff
@@ -1,6 +1,7 @@
-import { type ChangeEvent, useCallback, useState } from "react";
+import type { ChangeEvent } from "react";
+import { useCallback, useState } from "react";
```
why split imports from the same package?

akbarkz commented on static/client/pages/Webpage/Webpage.tsx:
```diff
@@ -135,9 +136,13 @@ const Webpage = ({ page, project }: IWebpageProps): ReactNode => {
-            <OwnerAndReviewers page={page} />
-            <div className="u-sv3" />
-            <Products page={page} />
+            <div className={isNew ? "col-12" : "col-5"}>
```
Can you use a new grid please?

akbarkz commented on static/client/pages/Webpage/Webpage.tsx:
```diff
@@ -153,11 +158,10 @@ const Webpage = ({ page, project }: IWebpageProps): ReactNode => {
           </div>
         ) : null}
-        <WebpageAssets page={page} />
+        <WebpageAssets projectName={page.project?.name || ""} url={page.url || ""} />
```
Maybe better to have default values for props in WebpageAssets?

akbarkz commented:
@muhammad-ali-pk Findings from QA:
1. When typing in someone's name into "Owner" or "Reviewers" field, their names come up as "name" instead of their real full names.
2. Unrelated to your PR: open 20-years page on ubuntu.com, the very first asset in the list shows a wrong file type.
3. Unrelated to your PR: when creating a new webpage, if you type in "Copy Doc" field with a valid copy doc URL, after creation it will still generate a new copy doc and lint to that, instead of the existing copy doc.

muhammad-ali-pk commented on static/client/components/Common/CustomSearchAndFilter/types.ts:
```diff
@@ -0,0 +1,17 @@
+  labelKey?: string;
+}
+
+export interface IUserSearchAndFilterProps extends ICustomSearchAndFilterProps<IUser> {}
```
It's not used anymore. Thanks for flagging it.

muhammad-ali-pk commented:
@akbarkz Appreciate the thorough review! I have addressed your code and QA review comments.

Miscellaneous code review comments addressed ([here](https://github.com/canonical/cs.canonical.com/pull/240/commits/fd6aa6c3f8e518c6bf68889a343b1b854330e036))

> Since you created a mock server, it would be nice if you added how to use it in the README as well

Detailed documentation added [here](https://github.com/canonical/cs.canonical.com/blob/f33c33a7d301efcfda316abb7126fab22ac20f8e/MOCK_SERVER_USAGE.md)

> If possible, could you please add comments (in github) in places where you applied each of 5 SOLID principles. That would be a good learning guide for us on concrete examples.

Since each single example spans over multiple files, it's not ideal to add comments pinpointing the exact location of an example. What I have done instead is, implement each principle in a separate commit and link those commits to the PR description. Please check PR description for more details on individual principle and linked examples.

> When typing in someone's name into "Owner" or "Reviewers" field, their names come up as "name" instead of their real full names.

Addressed [here](https://github.com/canonical/cs.canonical.com/pull/240/commits/8ca7be1338562f10938a82be517dfb1795b1def5)

> Unrelated to your PR: open 20-years page on ubuntu.com, the very first asset in the list shows a wrong file type.

Fixed [here](https://github.com/canonical/cs.canonical.com/pull/240/commits/2a8f3222f5c7e8b507d101a273816ae06c9b93d2)

> Unrelated to your PR: when creating a new webpage, if you type in "Copy Doc" field with a valid copy doc URL, after creation it will still generate a new copy doc and lint to that, instead of the existing copy doc.

Fixed [here](https://github.com/canonical/cs.canonical.com/pull/240/commits/cd09e80486b96952080dd46035ffe4011be6a488)

> I can't verify if that's a bug also on production, since creating a new page on production doesn't work right now. But create a new page and click on Save. The newly created page does not get highlighted in the sidebar on the left.

It was working for me, but maybe you had the "Project view" selected when creating a page? In any case, the new page creation should now select tree view if not already selected, and highlight the page. Please lemme know if it still doesn't work for you. Addressed [here](https://github.com/canonical/cs.canonical.com/pull/240/commits/ae745aabaa1c2aad86e7326ffa8177d1daa7f210)

akbarkz commented:
Thanks for fixing the issues Muhammad! I'm approving, but if you can have another look at the file type issue and fix it as well, that would be great.
<img width="357" height="308" alt="Screenshot From 2025-09-30 09-14-59" src="https://github.com/user-attachments/assets/f6a23930-0523-47c4-b110-3150ab6d8f90" />

muhammad-ali-pk commented:
> but if you can have another look at the file type issue and fix it as well,

@akbarkz As discussed on MM, the parser re-runs after 24 hours. We decreased the delay to 5 minutes to QA it, and the asset file type fix works. Though, we might have to remove the problematic asset manually from the prod DB.
